eventbus/kafka/sensor/kafka_sensor.go

```diff
@@ -0,0 +1,355 @@
+	KafkaTriggerHandler
+	depName string
+}
+
+func (t Triggers) List(event *cloudevents.Event) []*TriggerWithDepName {
+	triggers := []*TriggerWithDepName{}
+
+	for _, trigger := range t {
+		if depName, ok := trigger.DependsOn(event); ok {
+			triggers = append(triggers, &TriggerWithDepName{trigger, depName})
+		}
+	}
+
+	return triggers
+}
+
+func (t Triggers) Ready() bool {
+	for _, trigger := range t {
+		if !trigger.Ready() {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *KafkaSensor) Initialize() error {
+	config, err := s.Config()
+	if err != nil {
+		return err
+	}
+
+	// sensor specific config
+	config.Producer.Transaction.ID = s.hostname
+
+	client, err := sarama.NewClient(s.Brokers(), config)
+	if err != nil {
+		return err
+	}
+
+	consumer, err := sarama.NewConsumerGroupFromClient(s.groupName, client)
+	if err != nil {
+		return err
+	}
+
+	producer, err := sarama.NewAsyncProducerFromClient(client)
+	if err != nil {
+		return err
+	}
+
+	offsetManager, err := sarama.NewOffsetManagerFromClient(s.groupName, client)
+	if err != nil {
+		return err
+	}
+
+	s.client = client
+	s.consumer = consumer
+	s.kafkaHandler = &KafkaHandler{
+		Mutex:         &sync.Mutex{},
+		Logger:        s.Logger,
+		GroupName:     s.groupName,
+		Producer:      producer,
+		OffsetManager: offsetManager,
+		TriggerTopic:  s.topics.trigger,
+		Handlers: map[string]func(*sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()){
+			s.topics.event:   s.Event,
+			s.topics.trigger: s.Trigger,
+			s.topics.action:  s.Action,
+		},
+	}
+
+	return nil
+}
+
+func (s *KafkaSensor) Connect(ctx context.Context, triggerName string, depExpression string, dependencies []eventbuscommon.Dependency, atLeastOnce bool) (eventbuscommon.TriggerConnection, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	// connect only if disconnected, if ever the connection is lost
+	// the connected boolean will flip and the sensor listener will
+	// attempt to reconnect by invoking this function again
+	if !s.connected {
+		go s.Listen(ctx)
+		s.connected = true
+	}
+
+	if _, ok := s.triggers[triggerName]; !ok {
+		expr, err := govaluate.NewEvaluableExpression(strings.ReplaceAll(depExpression, "-", "\\-"))
+		if err != nil {
+			return nil, err
+		}
+
+		depMap := map[string]eventbuscommon.Dependency{}
+		for _, dep := range dependencies {
+			depMap[base.EventKey(dep.EventSourceName, dep.EventName)] = dep
+		}
+
+		s.triggers[triggerName] = &KafkaTriggerConnection{
+			KafkaConnection: base.NewKafkaConnection(s.Logger),
+			sensorName:      s.sensor.Name,
+			triggerName:     triggerName,
+			depExpression:   expr,
+			dependencies:    depMap,
+			atLeastOnce:     atLeastOnce,
+			close:           s.Close,
+			isClosed:        s.IsClosed,
+		}
+	}
+
+	return s.triggers[triggerName], nil
+}
+
+func (s *KafkaSensor) Listen(ctx context.Context) {
+	defer s.Disconnect()
+
+	for {
+		if len(s.triggers) != len(s.sensor.Spec.Triggers) || !s.triggers.Ready() {
```

Comment:
Is it possible that the triggers would never be ready and this would loop forever. Would it be helpful to add a max wait duration?

Reply:
Doesn't hurt, I can add one!

Reply:
So I implemented this and then I realized, even if we waited a max duration and return, the function would just be re-invoked in `sensors/listener.go` anyways. Furthermore, I realized that Connect and Subscribe are guaranteed to be called for each trigger (and if this happens then the if condition will be satisfied) so I don't think this is actually necessary.